instrumentation/src/main/java/io/opentelemetry/android/internal/services/periodicwork/PeriodicWorkService.kt

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.android.internal.services.periodicwork
+
+import android.os.Handler
+import android.os.Looper
+import io.opentelemetry.android.internal.services.Service
+import java.util.concurrent.LinkedBlockingQueue
+import java.util.concurrent.ThreadPoolExecutor
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicBoolean
+
+/**
+ * Utility to run periodic background work.
+ *
+ * <p>This class is internal and not for public use. Its APIs are unstable and can change at any
+ * time.
+ */
+class PeriodicWorkService : Service {
+    private val delegator = WorkerDelegator()
+    private val started = AtomicBoolean(false)
+
+    override fun start() {
+        if (!started.getAndSet(true)) {
+            delegator.run()
+        }
+    }
+
+    fun enqueue(runnable: Runnable) {
+        delegator.enqueue(runnable)
+    }
+
+    private class WorkerDelegator : Runnable {
+        companion object {
+            private const val SECONDS_TO_KILL_IDLE_THREADS = 30L
+            private const val SECONDS_FOR_NEXT_LOOP = 10L
+            private const val MAX_AMOUNT_OF_WORKER_THREADS = 1
+            private const val NUMBER_OF_PERMANENT_WORKER_THREADS = 0
+        }
+
+        private val queue = LinkedBlockingQueue<Runnable>()
+        private val handler = Handler(Looper.getMainLooper())
+        private val executor =
+            ThreadPoolExecutor(
+                NUMBER_OF_PERMANENT_WORKER_THREADS,
+                MAX_AMOUNT_OF_WORKER_THREADS,
+                SECONDS_TO_KILL_IDLE_THREADS,
+                TimeUnit.SECONDS,
+                LinkedBlockingQueue(),
+            )
+
+        fun enqueue(runnable: Runnable) =
+            synchronized(this) {
+                queue.add(runnable)
+            }
+
+        override fun run() {
+            delegateToWorkerThread()
+            scheduleNextLookUp()
+        }
+
+        private fun delegateToWorkerThread() =
+            synchronized(this) {
```

Comment:
This is causing contention on the `enqueue` method since it synchronizes the whole thing until the queue is empty (which might take a bit of time, I guess?).
Maybe an option is to use a `ConcurrentLinkedQueue` instead and just poll its items (as a copy) and run the while loop on the copy of the queue, so the queue is released and ready to be used again.

Reply:
This is a good point, since the `executor.execute()` function should just move the item to the executor's internal queue, I was thinking that the overhead shouldn't be too much. But it's definitely better to go extra safe for these kinds of scenarios, let me try it out 👍

Reply:
Thank you, I've made the changes and it all seems to be working well 👍 It's definitely better to avoid blocking ops whenever possible.